rust/kernel/sync/arc.rs

```diff
@@ -7,178 +7,216 @@
+        // pinned, and in `drop`. Both are compatible with the pin requirements.
+        init(pinned);
+
+        Ok(Ref {
+            ptr: NonNull::from(Box::leak(inner)),
+        })
     }
 
-    /// Determines if two reference-counted pointers point to the same underlying instance of `T`.
-    pub fn ptr_eq(a: &Self, b: &Self) -> bool {
-        core::ptr::eq(a.ptr.as_ptr(), b.ptr.as_ptr())
+    /// Deconstructs a [`Ref`] object into a usize.
+    ///
+    /// It can be reconstructed once via [`Ref::from_usize`].
+    pub fn into_usize(obj: Self) -> usize {
+        ManuallyDrop::new(obj).ptr.as_ptr() as _
     }
 
-    /// Deconstructs a [`Ref`] object into a raw pointer.
+    /// Borrows a [`Ref`] instance previously deconstructued via [`Ref::into_usize`].
+    ///
+    /// # Safety
     ///
-    /// It can be reconstructed once via [`Ref::from_raw`].
-    pub fn into_raw(obj: Self) -> *const T {
-        let no_drop = ManuallyDrop::new(obj);
-        no_drop.ptr.as_ptr()
+    /// `encoded` must have been returned by a previous call to [`Ref::into_usize`]. Additionally,
+    /// [`Ref::from_usize`] can only be called after *all* instances of [`RefBorrow`] have been
+    /// dropped.
+    pub unsafe fn borrow_usize(encoded: usize) -> RefBorrow<T> {
+        // SAFETY: By the safety requirement of this function, we know that `encoded` came from
+        // a previous call to `Ref::into_usize`.
+        let obj = ManuallyDrop::new(unsafe { Ref::from_usize(encoded) });
+
+        // SAFEY: The safety requirements ensure that the object remains alive for the lifetime of
+        // the returned value. There is no way to create mutable references to the object.
+        unsafe { RefBorrow::new(obj) }
     }
 
-    /// Recreates a [`Ref`] instance previously deconstructed via [`Ref::into_raw`].
+    /// Recreates a [`Ref`] instance previously deconstructed via [`Ref::into_usize`].
     ///
     /// # Safety
     ///
-    /// `ptr` must have been returned by a previous call to [`Ref::into_raw`]. Additionally, it
-    /// can only be called once for each previous call to [``Ref::into_raw`].
-    pub unsafe fn from_raw(ptr: *const T) -> Self {
+    /// `encoded` must have been returned by a previous call to [`Ref::into_usize`]. Additionally,
+    /// it can only be called once for each previous call to [``Ref::into_usize`].
+    pub unsafe fn from_usize(encoded: usize) -> Self {
         Ref {
-            ptr: NonNull::new(ptr as _).unwrap(),
+            ptr: NonNull::new(encoded as _).unwrap(),
         }
     }
 }
 
-impl<T: RefCounted + ?Sized> Deref for Ref<T> {
+impl<T: ?Sized> Ref<T> {
+    /// Determines if two reference-counted pointers point to the same underlying instance of `T`.
+    pub fn ptr_eq(a: &Self, b: &Self) -> bool {
+        core::ptr::eq(a.ptr.as_ptr(), b.ptr.as_ptr())
+    }
+
+    /// Returns a pinned version of a given `Ref` instance.
+    pub fn pinned(obj: Self) -> Pin<Self> {
```

Comment:
Yes, I use this in subsequent commits, for example, this [one](https://github.com/wedsonaf/linux/commit/15e0a8517db8763ffdf58d07d2a240828547bb69) where I remove `Arc` usage from the samples.

Reply:
Sounds good, let's keep it.
I have the feeling that, since `Ref<T>` is an intrinsically pinned type, `Pin<Ref<T>>` might be redundant. But it doesn't hurt to have it, and we can always address it in a separate PR.